--- Model_NoDef.py	2023-05-27 01:58:24.430386213 -0400
+++ build_dnn.py.updated	2023-05-27 01:20:07.249669892 -0400
@@ -1,9 +1,12 @@
+#!/usr/bin/env python3
+
 # DF model used for non-defended dataset
 from keras.models import Sequential
 from keras.layers import Conv1D, MaxPooling1D, BatchNormalization
 from keras.layers.core import Activation, Flatten, Dense, Dropout
-from keras.layers.advanced_activations import ELU
+from keras.layers import ELU
 from keras.initializers import glorot_uniform
+from keras.optimizers import Adamax
 
 class DFNet:
     @staticmethod
@@ -89,3 +92,11 @@
         model.add(Dense(classes, kernel_initializer=glorot_uniform(seed=0), name='fc3'))
         model.add(Activation('softmax', name="softmax"))
         return model
+
+def build(input_shape, nclasses):
+    model = DFNet.build(input_shape, nclasses)
+
+    opt = Adamax(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=1e-08, decay=0.0) # Optimizer
+    model.compile(loss="categorical_crossentropy", optimizer=opt,
+                metrics=["categorical_accuracy"])
+    return model
